File: 0001-two-sum/solution.cpp
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    // 最优解：一遍遍历哈希（先查后存，天然无自匹配问题）
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> seen;   // 值 -> 下标
        vector<int> res;
        for (int i = 0; i < nums.size(); i++) {
            int need = target - nums[i];            // 我需要的"搭档"值
            if (seen.find(need) != seen.end()) {    // 搭档已在前面出现过
                res.push_back(seen[need]);          // 更早的下标在前
                res.push_back(i);
                return res;
            }
            seen[nums[i]] = i;                      // 否则记下当前 (值, 下标)
        }
        return res;
    }

    /*
    // 替代解：两遍遍历哈希（先全存，再查；需注意自匹配）
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> mem;
        for (int i = 0; i < nums.size(); i++) mem[nums[i]] = i;
        for (int i = 0; i < nums.size(); i++) {
            int x = target - nums[i];
            if (mem.find(x) != mem.end() && mem[x] != i) {
                return {i, mem[x]};
            }
        }
        return {};
    }
    */
};

File: 0001-two-sum/solution.py
from typing import List

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        seen = {}                          # 值 -> 下标，哈希表
        for i, num in enumerate(nums):
            need = target - num            # 我需要的"搭档"值
            if need in seen:               # 搭档已经在前面出现过
                return [seen[need], i]     # 返回 [更早的下标, 当前下标]
            seen[num] = i                  # 否则把当前 (值, 下标) 记下来，供后面查
        return []                          # 题目保证有解，正常不会走到这里


# ---------- 验证（本地跑，提交时这段可留可删）----------
if __name__ == "__main__":
    s = Solution()
    print(s.twoSum([2, 7, 11, 15], 9))    # [0, 1]
    print(s.twoSum([3, 2, 4], 6))         # [1, 2]
    print(s.twoSum([3, 3], 6))            # [0, 1]
